File: control-plane/src/actions.ts
// Action catalog (control-plane view).
//
// The AGENT is the authoritative validator and the only place that maps an
// action to concrete commands (see agent/internal/actions). The control plane
// keeps this lightweight catalog so it can:
//   - reject enqueue of an unknown action early, and
//   - know whether an action is idempotent (drives interrupted-job redelivery).
//
// Keep this list in sync with PROTOCOL.md §7 and the Go action registry.

export interface ActionSpec {
  name: string;
  idempotent: boolean;
  /** Lightweight shape check; the agent performs strict validation. */
  validate(params: Record<string, unknown>): string | null; // null = ok
}

const BRANCH_RE = /^[A-Za-z0-9._/-]{1,255}$/;
const APP_RE = /^[A-Za-z0-9._-]{1,64}$/;
const HOOK_RE = /^[a-z0-9][a-z0-9_-]{0,63}$/;

export const ACTIONS: Record<string, ActionSpec> = {
  healthcheck: {
    name: "healthcheck",
    idempotent: true,
    validate: () => null,
  },
  deploy: {
    name: "deploy",
    idempotent: false,
    validate: (p) => {
      const branch = p.branch;
      if (branch === undefined) return null;
      if (typeof branch !== "string" || !BRANCH_RE.test(branch))
        return "invalid branch";
      if (branch.includes("..") || branch.startsWith("-"))
        return "invalid branch";
      return null;
    },
  },
  restart: {
    name: "restart",
    idempotent: true,
    validate: (p) => {
      const app = p.app;
      if (app === undefined) return null;
      if (typeof app !== "string" || !APP_RE.test(app)) return "invalid app";
      return null;
    },
  },
  "run-hook": {
    name: "run-hook",
    idempotent: false,
    validate: (p) => {
      const name = p.name;
      if (typeof name !== "string" || !HOOK_RE.test(name)) return "invalid hook name";
      return null;
    },
  },
};

export function isKnownAction(name: string): boolean {
  return Object.prototype.hasOwnProperty.call(ACTIONS, name);
}

File: agent/internal/actions/actions.go
// Package actions implements the agent's allowlist of named actions.
//
// SECURITY MODEL: this is the only place that maps an action name to concrete
// commands. The API never supplies a command string — it sends a
// named action plus a constrained params object. Adding an action is a code
// change, reviewable in a PR. Unknown actions and invalid params are refused.
package actions

import (
	"context"
	"encoding/json"
	"fmt"
)

// LogFunc streams a chunk of output for a stream ("stdout" | "stderr").
// The client layer assigns sequence numbers and frames it as a protocol Log.
type LogFunc func(stream, chunk string)

// Handler executes an action. It returns the process exit code (0 == success)
// and an optional error for agent-level failures. Honor ctx for cancellation
// and timeouts.
type Handler func(ctx context.Context, params json.RawMessage, emit LogFunc) (int, error)

// Action is an allowlisted, named unit of work.
type Action struct {
	Name string
	// Idempotent marks actions that are safe to re-run after an interrupted
	// dispatch (drives redelivery semantics in the API).
	Idempotent bool
	Handle     Handler
}

// Registry is the closed set of supported actions. Keep in sync with
// PROTOCOL.md §7 and api/src/actions.ts.
var Registry = map[string]Action{
	"healthcheck": {Name: "healthcheck", Idempotent: true, Handle: handleHealthcheck},
	"deploy":      {Name: "deploy", Idempotent: false, Handle: handleDeploy},
	"restart":     {Name: "restart", Idempotent: true, Handle: handleRestart},
	"run-hook":    {Name: "run-hook", Idempotent: false, Handle: handleRunHook},
}

// Names returns the sorted-stable allowlist for the hello message.
func Names() []string {
	out := make([]string, 0, len(Registry))
	for _, name := range []string{"healthcheck", "deploy", "restart", "run-hook"} {
		if _, ok := Registry[name]; ok {
			out = append(out, name)
		}
	}
	return out
}

// Lookup returns the action and whether it is in the allowlist.
func Lookup(name string) (Action, bool) {
	a, ok := Registry[name]
	return a, ok
}

// ErrRefused indicates the action or its params were rejected (exit 126).
type ErrRefused struct{ Reason string }

func (e ErrRefused) Error() string { return e.Reason }

func refuse(format string, args ...any) error {
	return ErrRefused{Reason: fmt.Sprintf(format, args...)}
}
